tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/cluster/distribution/ClusterNotificationsTest.java

```diff
@@ -0,0 +1,124 @@
+   }
+
+   @Test
+   public void testClusterNotificationsContinuity() throws Exception {
+      final MessageLoadBalancingType lbType = MessageLoadBalancingType.OFF_WITH_REDISTRIBUTION;
+      final String namePrefix = "TEST.QUEUE.";
+      final int queueCount = 500;
+
+      AssertionLoggerHandler loggerHandler = new AssertionLoggerHandler();
+      runAfter(loggerHandler::close);
+
+      setupServer(0, isFileStorage(), isNetty());
+      setupServer(1, isFileStorage(), isNetty());
+      setupServer(2, isFileStorage(), isNetty());
+
+      setupClusterConnection("cluster0", "", lbType, 1, isNetty(), 0, 1, 2);
+      setupClusterConnection("cluster0", "", lbType, 1, isNetty(), 1, 2, 0);
+      setupClusterConnection("cluster0", "", lbType, 1, isNetty(), 2, 0, 1);
+
+      startServers(0, 1, 2);
+
+      AddressSettings as = new AddressSettings().setRedistributionDelay(0);
+      servers[0].getAddressSettingsRepository().addMatch(namePrefix + "#", as);
+      servers[1].getAddressSettingsRepository().addMatch(namePrefix + "#", as);
+      servers[2].getAddressSettingsRepository().addMatch(namePrefix + "#", as);
+
+      waitForTopology(servers[2], 3);
+
+      ExecutorService executorService = Executors.newCachedThreadPool();
+      CountDownLatch latch = new CountDownLatch(queueCount * 2);
+
+      try (ActiveMQConnectionFactory factory = new ActiveMQConnectionFactory("tcp://localhost:61616")) {
+
+         List<DefaultMessageListenerContainer> containers = new ArrayList<>();
+         runAfter(() -> containers.parallelStream().forEach(AbstractJmsListeningContainer::shutdown));
+         runAfter(() -> executorService.shutdown());
+
+         for (int i = 0; i < queueCount; i++) {
+            Queue queue = ActiveMQDestination.createQueue(namePrefix + i + "-");
+
+            executorService.submit(() -> {
+               try (Connection connection = factory.createConnection();
+                    Session session = connection.createSession(Session.SESSION_TRANSACTED)) {
+
+                  session.createProducer(queue).send(session.createTextMessage("Message"));
+                  session.commit();
+                  latch.countDown();
+               } catch (Exception e) { }
+            });
+
+            DefaultMessageListenerContainer container = new DefaultMessageListenerContainer();
+            container.setCacheLevelName("CACHE_NONE");
+            container.setSessionTransacted(true);
+            container.setSessionAcknowledgeModeName("SESSION_TRANSACTED");
+            container.setConcurrentConsumers(20);
+            container.setConnectionFactory(new CachingConnectionFactory(factory));
+            container.setDestinationName(queue.getQueueName());
+            container.setBackOff(new FixedBackOff(1000, 2));
+            container.setReceiveTimeout(100);
+            container.setMessageListener((MessageListener) msg -> {
+               latch.countDown();
+            });
+
+            container.initialize();
```

Comment:
This test is always failing for me:


```

java.lang.OutOfMemoryError: unable to create native thread: possibly out of memory or process/resource limits reached

	at java.base/java.lang.Thread.start0(Native Method)
	at java.base/java.lang.Thread.start(Thread.java:802)
	at org.springframework.core.task.SimpleAsyncTaskExecutor.doExecute(SimpleAsyncTaskExecutor.java:248)
	at org.springframework.core.task.SimpleAsyncTaskExecutor.execute(SimpleAsyncTaskExecutor.java:202)
	at org.springframework.core.task.SimpleAsyncTaskExecutor.execute(SimpleAsyncTaskExecutor.java:180)
	at org.springframework.jms.listener.DefaultMessageListenerContainer.doRescheduleTask(DefaultMessageListenerContainer.java:780)
	at org.springframework.jms.listener.AbstractJmsListeningContainer.resumePausedTasks(AbstractJmsListeningContainer.java:545)
	at org.springframework.jms.listener.AbstractJmsListeningContainer.doStart(AbstractJmsListeningContainer.java:299)
	at org.springframework.jms.listener.AbstractJmsListeningContainer.start(AbstractJmsListeningContainer.java:277)
	at org.springframework.jms.listener.DefaultMessageListenerContainer.start(DefaultMessageListenerContainer.java:657)
	at org.apache.activemq.artemis.tests.integration.cluster.distribution.ClusterNotificationsTest.testClusterNotificationsContinuity(ClusterNotificationsTest.java:110)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:232)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:55)

```


Can we do this test without spring though? it gets easier to debug tests. I'm not aware of any other tests actually using spring for consuming messages and asserting.


I reckon it removes a lot of boilerplate code from the consumer.. but I would prefer having more control on what's going on... 


like now: the tests Is failing with lack of threads and I have no idea how to fix it.

Reply:
Yes, I would also _really_ prefer to do this without Spring but so far I have been unable to reproduce this behavior without Spring (outside of production of course).

Closest thing I got to working involved "rolling restarts" of all nodes in a cluster with the connected clients running failover-semantics... this caused the issue a few times but not consistently _at all_ and the overall setup was to chaotic to make anything useful of it regardless.

Reply:
on the case of the resources.. I think this test should be moved to tests/soak-tests... 

some tests I have are using parameters.sh... and we could then increase the sizing...

We should be able to run the tests on our envs.. and that kind of variable be applied to make it use a bigger box

Reply:
Yeah, that makes sense, I will move the test tomorrow.

Where you able to run the current test to trigger the issue by reducing concurrentConsumers and queueCount? Wondering if it's strictly timing plus resources that makes the difference or something more...

Reply:
FWIW, the test runs fine on my machine. I was able to back out the fix, reproduce the issue, change some stuff, retest, etc.

Reply:
Ok.  Let's make it a soak test. And merge this PR. 

@AntonRoskvist let me know if you have any issues doing it.  If you can't let me know.